### DISCUSSION TITLE
How to add "implements" contracts to drift defined views?

### DISCUSSION THREAD

**Discussion by Adam-Langley** [simolus3/drift#2754] [Q&A] [answered]
I have views defined in .drift files, and I want the data-classes (and the ViewInfo) to implement my custom interfaces (numerous).
Can I achieve this?

Here is the use-case.

I have several tables, used to populate dropdown lookups. They are enhanced/supplemented using some (irrelevant) SQl logic - hence being views, rather than directly accessing the table.
Because they are lookups, they have sorting defined - lets call the column "sort_order" type int.

I want to define a custom abstract classes
abstract class WithSortOrder
{
  int sortOrder;
}
abstract class WithSortOrderField
{
 late final GeneratedColumn<int> sortOrder = GeneratedColumn<int>(
      'sort_order', aliasedName, false,
      type: DriftSqlType.int);
}
These will let me sort the rows both during, and after retrieving them from the database (in an abstracted manner, via SQL or dart)

And I want the code generated by drift to apply the appropriate "implements WithSortOrder" to my view DataClass and ViewInfo.

This will let me centralise the sorting in my repositories, such that I can always be sure my lists will be ordered correctly as defined in the database.

The example below uses 2 custom contracts, for multiple sorting.
```
  query.orderBy([
      if (t is WithIsDefaultField)
        (u) =>
          OrderingTerm(expression: (u as WithIsDefaultField).isDefault, mode: OrderingMode.desc),

      if (t is WithSortOrderField)
        (u) =>
          OrderingTerm(expression: (u as WithSortOrderField).orderIndex, mode: OrderingMode.asc)
    ]);
```

Summary:
I want to be able to define multiple abstract classes, and somehow get the generated code (both the DataClass and the ViewInfo<>) to declare that they "implement" them.

Is this possible?

Thank you

**Comment by simolus3** [comment]
It's not possible to specify a base class for tables and views declared in drift files - neither for the data class nor for the table/view info class.

It's not quite as reliable, but if no other table has columns with those names you could look them up via [runtime schema inspection](https://drift.simonbinder.eu/docs/dart-api/schema_inspection/):

```dart
  final isDefault = t.columnsByName['is_default'];
  final sortOrder = t.columnsByName['sort_order'];

  query.orderBy([
      if (isDefault != null)
        (u) =>
          OrderingTerm(expression: isDefault, mode: OrderingMode.desc),

      if (sortOrder != null)
        (u) =>
          OrderingTerm(expression: sortOrder, mode: OrderingMode.asc)
    ]);
```

**Comment by Adam-Langley** [comment, marked as answer]
Thanks @simolus3 for that suggestion.
I did actually find a way to achieve this:

Write a custom class which derives from the generated view - its sole purpose is to declare that the view implements certain contracts:

```
/// MyCustomDriftDslView is the view class generated by the .drift view sql
class MyCustomDriftDslViewWithContracts
    extends MyCustomDriftDslView
    implements WithSortOrderField, WithIsDefaultField {
  MyCustomDriftDslViewWithContracts(super.attachedDatabase);
}
```

Then manually attach the view to the database during construction:
```
class MyDatabase extends _$MyDatabase {
  MyDatabase(LazyDatabase lazy) : super(lazy);

  MyDatabase.from([QueryExecutor? executor]) : super(executor!);

  late final MyCustomDriftDslViewWithContracts myCustomView = MyCustomDriftDslViewWithContracts(this);
}
```
I guess the most significant difference to what you have suggested, would be this solution will apply compile-time checking.

Thanks for the tip - now I have a couple of tools to achieve what I want.
Thank you
